src/transaction_builder.cpp

```diff
@@ -264,6 +287,34 @@ void TransactionBuilder::SetExpiryHeight(uint32_t nExpiryHeight)
     mtx.nExpiryHeight = nExpiryHeight;
 }
 
+bool TransactionBuilder::AddOrchardSpend(
+    libzcash::OrchardSpendingKey sk,
+    orchard::SpendInfo spendInfo)
+{
+    if (!orchardBuilder.has_value()) {
+        // Try to give a useful error.
+        if (!(jsInputs.empty() && jsOutputs.empty())) {
+            throw std::runtime_error("TransactionBuilder cannot spend Orchard notes in a Sprout transaction");
+        } else if (mtx.nVersion < ZIP225_MIN_TX_VERSION) {
+            throw std::runtime_error("TransactionBuilder cannot spend Orchard notes before NU5 activation");
+        } else {
+            throw std::runtime_error("TransactionBuilder cannot spend Orchard notes without Orchard anchor");
+        }
+    }
+
+    auto fromAddr = spendInfo.FromAddress();
+    auto value = spendInfo.Value();
+    auto res = orchardBuilder.value().AddSpend(std::move(spendInfo));
+    if (res) {
+        orchardSpendingKeys.push_back(sk);
+        if (!firstOrchardSpendAddr.has_value()) {
```

Comment:
nonblocking: I'm a little surprised to see the continued primacy of the address as the "thing we're spending from" rather than the account or the IVK. Since the notes are received at specific addresses this is fine for now, and I guess that since we're in the transaction builder it's possible that we're being called from a context where we're spending notes from multiple accounts, like in z_mergetoaddress. Maybe in the future we can revisit the "source of funds" question to make it more explicit; depending upon the order of sources has always seemed like a bit of a smell to me.

Reply:
Sure, we can revisit this later, but right now this is the information I have available to work with when the caller doesn't set an explicit change address. The caller is welcome to do so, and in fact they are the best place to handle any account etc information as they know where their notes came from, whereas the transaction builder just sees a set of notes from arbitrary spending keys.